Prevent the details sidebar from auto close when IOU cancellable data is available

jayeshmangwani opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

### Fixed Issues
<!---
1. Please replace GH_LINK with a URL link to the GitHub issue this Pull Request is fixing.
2. Please replace PROPOSAL: GH_LINK_ISSUE(COMMENT) with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>
$ https://github.com/Expensify/App/issues/<number-of-the-issue(comment)>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/App/issues/16508
PROPOSAL: https://github.com/Expensify/App/issues/16508#issuecomment-1527025599


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

**Case 1: Cancel multiple requests**

1. Open App and navigate to chat
2. Send multiple money requests to user B from user A
3. Open the details sidebar screen
3. Cancel requests one by one from user A by pressing the cancel button
4. Verify that the sidebar remains Open until the last outstanding request

**Case 2: Reject multiple requests**
1. Go to user B
2. Send multiple money requests to user A
3. Open the details sidebar screen
3. Go to user A and Decline the request one by one by pressing the decline button
4. Verify that the sidebar remains Open until the last outstanding request

**Case 3: Settle up  button**
1. Go to user B
2. Send multiple money requests to user A
3. Open the details sidebar screen
3. Go to user A and press I'll settle up elsewhere
4. Verify that the sidebar closes

**Case 4: Cancel a request**
1. Open App and navigate to chat
2. Send a money request to user B from user A
3. From user A, Open the details sidebar screen
3. Cancel the request by pressing the cancel button
4. Verify that the sidebar closes 

**Case 5: Reject a request**
1. Go to user B
2. Send a money request to user A
3. Go to user A,  Open the details sidebar screen, and Decline the request by pressing the decline/ I'll settle up elsewhere button
4. Verify that the sidebar closes 

- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->
Same as Tests

<!---
1. Open App and navigate to chat
2. Go offline
3. Send multiple money requests to user B from user A
4. Cancel requests one by one from user A
5. Verify that the sidebar remains Open until the last outstanding request, Go online on user A
1. Go to user B
2. Send multiple money requests to user A
3. Go to user A
4. Go offline
5. Decline the request one by one
6. Verify that the sidebar remains Open until the last outstanding request
-->

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->
Same as Tests

- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/35371050/235299777-d191ddf4-6e86-44d0-9316-89d8d277601d.mp4


<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/35371050/235299784-81693705-6dcf-4c89-8870-97bbf7077e8f.mp4


<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/35371050/235299787-cbcc937a-36ff-4f9b-a812-a251988acba6.mp4


<!-- add screenshots or videos here -->

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/35371050/235299791-3a0a1ef4-be58-4a59-a156-c2abb62cd6fa.mp4


<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/35371050/235299796-78ae2362-8bcc-4166-a409-4ea3ad7023d6.mov


<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/35371050/235299800-36ec7b92-e7ed-42d1-8fce-8f03db00d22f.mp4


<!-- add screenshots or videos here -->

</details>

jayeshmangwani committed:
feat: added a condition if we should call navigate to CancelMoneyRequest

jayeshmangwani committed:
feat: added a new prop to cancelMoneyRequest to handle navigation for last cancel request

jayeshmangwani committed:
feat: when rejectableTransactions length will one then close sidebar

jayeshmangwani committed:
Merge branch 'main' into details_sidebar_auto_close

jayeshmangwani committed:
Merge branch 'main' into details_sidebar_auto_close

Julesssss reviewed (commented):
Can we add a test case for non-reject actions too please? Maybe just this: 

- reject all payments, should close
- create new request
- pay
- it should close (I think this is true)

jayeshmangwani commented:
@Julesssss  I added a few different test cases, please check if this will work or if we need to modify them, and I have updated the Tests section only, and if these tests will work, then should we need to add all the cases offline Tests section too?

jayeshmangwani committed:
Merge branch 'main' into details_sidebar_auto_close

Julesssss commented:
Thanks, that looks good to me. 

> and if these tests will work, then should we need to add all the cases offline Tests section too?

Yeah, lets just adda. comment in offline tests that all of the above should be tested.

jayeshmangwani committed:
Merge branch 'main' into details_sidebar_auto_close

jayeshmangwani commented:
@rushatgabhane bump for review

jayeshmangwani committed:
Merge branch 'main' of https://github.com/Expensify/App into details_sidebar_auto_close

Julesssss reviewed (dismissed):
Just waiting for the full C+ checklist. Looking good 👍

amyevans commented on src/components/ReportTransaction.js:
```diff
@@ -30,11 +30,15 @@ const propTypes = {
     /** Type of the reject transaction button */
     rejectButtonType: PropTypes.oneOf([CONST.IOU.REPORT_ACTION_TYPE.DECLINE, CONST.IOU.REPORT_ACTION_TYPE.CANCEL]).isRequired,
 
+    /** Indicates whether the pressing reject button we should hide the details sidebar */
```
NAB, just a suggestion to improve the comment's clarity
```suggestion
    /** Indicates whether pressing the reject button should hide the details sidebar */
```

jayeshmangwani replied:
@amyevans comment updated

amyevans reviewed (dismissed):
Looks good, thank you for the clear outline of test cases!

jayeshmangwani committed:
Merge branch 'main' into details_sidebar_auto_close

jayeshmangwani committed:
fix: comments updated to be more clarified

rushatgabhane reviewed (approved):
@amyevans LGTM too

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

luacmartins commented:
Just dropping a note that I don't think this feature is needed post Manual Requests. We don't have the sidebar modal anymore. I resolved some conflicts with my PR and left the feature in, but we'll likely remove it soon.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/4922047640) to production by https://github.com/roryabraham in version: 1.3.12-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
